src/v/wasm/ffi.h

```diff
@@ -222,7 +228,9 @@ void transform_type(std::vector<val_type>& types) {
     } else if constexpr (
       std::is_same_v<Type, int64_t> || std::is_same_v<Type, uint64_t>) {
         types.push_back(val_type::i64);
-    } else if constexpr (std::is_pointer_v<Type> || std::is_integral_v<Type>) {
+    } else if constexpr (
+      std::is_pointer_v<Type> || std::is_integral_v<Type>
+      || std::is_same_v<Type, ptr>) {
         types.push_back(val_type::i32);
```

Comment:
question: guest_ptr is `uint32` but here we register it as `i32` which implies signed. Will that be a problem down the line?

Reply:
Good question! It's a little weird in Wasm, but types aren't signed only the operations that act upon them (sort of like how register and the CPU work), so this is really passing 32 bits and it's up to us to interpret that as signed or unsigned.

Reply:
For more see the spec here:
https://webassembly.github.io/spec/core/syntax/types.html

Inlined the relevant bit here:

> The types and classify 32 and 64 bit integers, respectively. Integers are not inherently signed or unsigned, their interpretation is determined by individual operations.

Reply:
That's interesting and a good read. I suppose this means that we will have to check for signed/unsigned at some point?

Reply:
No you just need to cast.